### PR TITLE
Scalastyle & compiler config updates

### PR DESCRIPTION
**Pull request by akkomar**
In this PR:
* scalastyle config is cleaned and nicely commented
* sbt version is bumped
* compiler is configured to:
  * warn on unused imports and declarations
  * fail the build on any warnings

Scalastyle-related commits should be squashed before merging.

### CODE REVIEW

File: scalastyle-config.xml [hunk 1]
```diff
@@ -48,15 +48,9 @@
...
-    <check level="error" class="org.scalastyle.scalariform.MagicNumberChecker" enabled="false">
-        <parameters>
-            <parameter name="ignore"><![CDATA[-1,0,1,2,3]]></parameter>
-        </parameters>
-    </check>
     <check level="error" class="org.scalastyle.scalariform.NoWhitespaceBeforeLeftBracketChecker" enabled="true"></check>
     <check level="error" class="org.scalastyle.scalariform.NoWhitespaceAfterLeftBracketChecker" enabled="true"></check>
-    <check level="error" class="org.scalastyle.scalariform.ReturnChecker" enabled="true"></check>
-    <check level="error" class="org.scalastyle.scalariform.NullChecker" enabled="false"></check>
+    <check customId="return" level="error" class="org.scalastyle.scalariform.ReturnChecker" enabled="true"></check>
```
Comment (mreid-moz) [inline scalastyle-config.xml]: why do we need a `customId` property here?

File: scalastyle-config.xml [hunk 2]
```diff
@@ -101,5 +90,38 @@
...
+    <check level="warn" class="org.scalastyle.scalariform.IllegalImportsChecker" enabled="true">
+        <parameters>
+            <parameter name="illegalImports"><![CDATA[org.joda._]]></parameter>
+        </parameters>
+    </check>
+
+    <!--=============================================================================================================-->
+    <!--                                     rules we don't want to enforce yet                                      -->
+    <!--=============================================================================================================-->
+    <check enabled="false" class="org.scalastyle.scalariform.UnderscoreImportChecker" level="warning">
```
Comment (mreid-moz) [inline scalastyle-config.xml]: nit: other rules have the `level` property first and `enabled` at the end - consider keeping properties in the same order throughout.

Comment (fbertsch) [reply scalastyle-config.xml]: Do we need an XML style checker? 😁

Comment (akkomar) [reply scalastyle-config.xml]: Yes, but we'll use json for defining rules.

File: scalastyle-config.xml (hunk 1 above)
Comment (akkomar) [reply scalastyle-config.xml]: For disabling this check, which would be helpful for t-b-v.
We have some complex functions there that would require refactoring in order to pass this - I'd opt for disabling this check in some places at first.

**Pull request merged**